rfc/0015-multi-file-jayvee/0015-multi-file-jayvee.md

```diff
@@ -0,0 +1,116 @@
+
+
+## Motivation
+
+Right now, Jayvee users can only pack their whole Jayvee model into one file. 
+The challenge is two-fold:
+1. Larger projects will become unmaintainable quite quickly, as the elements cannot be organized into multiple files.
+2. Without distribution to multiple files, there is no possibility to reuse models of other projects.
+
+For example, we might be enable to build libraries of valuetypes that can be reused across multiple projects instead of copying the code.
+
+## Explanation
+
+### Exporting elements
+
+For exporting elements, I propose introducing a new concept called `libraries` instead of explicitly modeling an import or export per element.
+A `library` can inhibit `Valuetype`s, `Block`s, `BlockType`s,  `Constraint`s, and `Transform`s.
+
+**Example library**
+```
+library MyDomainLibrary {
+  valuetype MyDomainSpecificValuetype {
+    // ... details of valuetype
+  }
+
+  // ... possibly more elements
+}
+```
+
+Libraries are always "exported" and can be "imported" into other files.
+
+
+### Visibility of elements in a file
+
+By introducing the concept of `libraries`, most elements can be defined on three levels in a Jayvee file:
+1. On the root level of the file
+2. Within a pipeline
+3. (new) Within a library
+
+The name of an element is given by its definition. The **qualified name** is constructed by prepending container structures in this pattern: `<container name>.<element name>`, e.g., `MyDomainLibraryMyDomainLibrary.MyDomainSpecificValuetype`.
+
+**Access paths:**
+- root level elements can access 
+  - root level elements by their name
+  - and elements of libraries by their qualified name
+- elements within a library can access
+  - root level elements by their name
+  - elements within the same library by their name
+  - and elements of other libraries by their qualified name
+- elements within a pipeline can access
+  - root level elements by their name
+  - elements within the same pipeline by their name
+  - and elements of libraries by their qualified name
+
+**No-access paths:**
+- elements within a pipeline cannot be referenced by outside elements
+
+
+### Importing elements
+
+Only `libraries` and their elements can be imported into other files. Elements on the root level of a file or within a pipeline cannot be imported.
+
+```
+import './path/to/jv-file.jv'; // imports all libraries of the file
+import './path/to/jv-file.jv' use { 
+  MyDomainLibrary 
+}; // only imports the named library, access via qualified name
+import './path/to/jv-file.jv' use {
+  MyDomainLibrary called MyLibraryAlias
+}; // only imports the named library, access via qualified name using the alias
+import './path/to/jv-file.jv' use {
+  MyDomainLibrary.MyDomainSpecificValuetype called MyElementAlias
+}; // only imports the named element, access via the alias
+```
+
+References to these imported elements is by their qualified name (unless altered by an alias).
+
+## Drawbacks
```

Comment:
This is missing no way to version libraries. I'd consider this either:

- part of the location string
- explicit in the import, e.g. `from 'location' version 3.1.2 use { Name }` with the version string/no string being enforced semver. Could also force the library to define it's own version.

Reply:
I like the idea!

Enforcing it in the location string might be weird for local libraries I guess.
The explicit import syntax suggests there is a version per file. I think we should have rather a version per library (potentially multiple in a file). Thus, I suggest to use the syntax `from './path/to/location.jv' use { MyDomainLibrary version 1.2.3 };`

Reply:
Hmm, I think that brings up the question if we want to import files and libraries the same. I like the explicit source/version etc import for libraries a lot, for files I think it is a bit of overkill. Maybe we could have a concept of including files and importing libraries?

Reply:
Right now, the RFC only covers importing libraries. Do we need to import a whole file at all?

Reply:
It makes sense, yeah. But that has a few implications, like having the visibility keyword at the root level as well, right? Or would you circumvent that?

Having two mechanisms in place allows us to be more specific, but also puts more load on the user. Especially if things are quite close together, that might be very confusing.
What if we remodel this RFC to only support local multi-file?
A future RFC would be to use `libraries` as an entrypoint to a versioned Jayvee model for reuse only for remote location import? That is also the case where we need versioning, for local import/export not so much.

Reply:
Hmm ya I think local file import first, then libraries would be good. Smaller and easier to execute on.

I like having visibility only in libraries, that way normal users don't need to use libraries (so have no visibility, no version, nothing, just their pipeline). And advanced users / library maintainers can deal with the overhead of maintaining a library if they actually want to do that.

Reply:
Added it to the draft.